core/src/banking_stage/thread_aware_account_locks.rs

```diff
@@ -0,0 +1,692 @@
+            let thread_id = thread_selector(schedulable_threads);
+            self.lock_accounts(write_account_locks, read_account_locks, thread_id);
+            thread_id
+        })
+    }
+
+    /// Returns `ThreadSet` that the given accounts can be scheduled on.
+    fn accounts_schedulable_threads<'a>(
+        &self,
+        write_account_locks: impl Iterator<Item = &'a Pubkey>,
+        read_account_locks: impl Iterator<Item = &'a Pubkey>,
+    ) -> ThreadSet {
+        let mut schedulable_threads = ThreadSet::any(self.num_threads);
+
+        for account in write_account_locks {
+            schedulable_threads &= self.write_schedulable_threads(account);
+        }
+
+        for account in read_account_locks {
+            schedulable_threads &= self.read_schedulable_threads(account);
+        }
+
+        schedulable_threads
+    }
+
+    /// Returns `ThreadSet` of schedulable threads for the given readable account.
+    /// If the account is not locked, then all threads are schedulable.
+    /// If only read locked, then all threads are schedulable.
+    /// If write-locked, then only the thread holding the write lock is schedulable.
+    /// The sequential limit is checked, and a thread will not be returned as schedulable
+    /// if the limit is reached.
+    fn read_schedulable_threads(&self, account: &Pubkey) -> ThreadSet {
+        // If the account is only read locked, then a read lock can be taken on any thread
+        // that is not at the sequential limit.
+        self.schedulable_threads_with_read_only_handler(account, |thread_set, counts| {
+            let mut schedulable_threads = ThreadSet::any(self.num_threads);
+            for thread_id in thread_set.threads_iter() {
+                if counts[thread_id] == self.sequential_queue_limit {
+                    schedulable_threads.remove(thread_id);
+                }
+            }
+            schedulable_threads
+        })
+    }
+
+    /// Returns `ThreadSet` of schedulable threads for the given writable account.
+    /// If the account is not locked, then all threads are schedulable.
+    /// If read-locked on a single thread, then only that thread is schedulable.
+    /// If write-locked, then only that thread is schedulable.
+    /// In all other cases, no threads are schedulable.
+    /// The sequential limit is checked, and a thread will not be returned as schedulable
+    /// if the limit is reached.
+    fn write_schedulable_threads(&self, account: &Pubkey) -> ThreadSet {
+        // If the account is only read locked, then a write lock can only be taken
+        // if the read lock is held by a single thread, and the limit is not exceeded.
+        self.schedulable_threads_with_read_only_handler(account, |thread_set, counts| {
+            thread_set
+                .only_one_scheduled()
+                .filter(|thread_id| counts[*thread_id] < self.sequential_queue_limit)
+                .map_or_else(ThreadSet::none, ThreadSet::only)
```

Comment:
hmm, another subjective cleaning:

```patch
         self.schedulable_threads_with_read_only_handler(account, |thread_set, counts| {
             thread_set
                 .only_one_scheduled()
-                .filter(|thread_id| counts[*thread_id] < self.sequential_queue_limit)
-                .map_or_else(ThreadSet::none, ThreadSet::only)
+                .and_then(|thread_id| (counts[thread_id] < self.sequential_queue_limit).then(|| ThreadSet::only(thread_id)))
+                .unwrap_or(ThreadSet::none())
         })
     }

```

i.e. `map_or_else` is hard to recognize its arg order.

Reply:
that's fair, I also hate the ordering of the args to `map_or_else` - my mental model is always reversed: (map, or_else), rather than what it is (or_else, map).

Reply:
I used a map + unwrap_or_else: d1ef4e0c8963702c761bd19ec480418f788a2684 (move away from `map_or_else`) + 6a8a0a89fd20d3a91ec936828df6e0052dbf0fed (filter removed)